crates/node/src/kitchensink.rs

```diff
@@ -0,0 +1,506 @@
+        self.process_proxy = Some(proxy_process);
+
+        Ok(())
+    }
+
+    fn generate_chainspec_file(&self, genesis_path: &str) -> anyhow::Result<PathBuf> {
+        let mut balances = self.extract_balance_from_genesis_file(genesis_path)?;
+
+        let mut chainspec_json: JsonValue =
+            serde_json::from_str(include_str!("default_chainspec.json"))?;
+
+        let existing_balances =
+            chainspec_json["genesis"]["runtimeGenesis"]["patch"]["balances"]["balances"]
+                .as_array()
+                .cloned()
+                .unwrap_or_default();
+
+        let mut merged_balances: Vec<(String, u128)> = existing_balances
+            .into_iter()
+            .filter_map(|val| {
+                if let Some(arr) = val.as_array() {
+                    if arr.len() == 2 {
+                        let account = arr[0].as_str()?.to_string();
+                        let balance = arr[1].as_f64()? as u128;
+                        return Some((account, balance));
+                    }
+                }
+                None
+            })
+            .collect();
+
+        merged_balances.append(&mut balances);
+
+        chainspec_json["genesis"]["runtimeGenesis"]["patch"]["balances"]["balances"] =
+            json!(merged_balances);
+
+        let temp_path =
+            std::env::temp_dir().join(format!("chainspec-{}.json", uuid::Uuid::new_v4()));
+
+        let mut temp_file = std::fs::File::create(&temp_path)?;
+        serde_json::to_writer_pretty(&mut temp_file, &chainspec_json)?;
+        temp_file.flush()?;
+
+        Ok(temp_path)
+    }
+
+    fn extract_balance_from_genesis_file(
+        &self,
+        genesis_path: &str,
+    ) -> anyhow::Result<Vec<(String, u128)>> {
+        let genesis_str = fs::read_to_string(genesis_path)?;
+        let genesis_json: Value = serde_json::from_str(&genesis_str)?;
+        let alloc = genesis_json
+            .get("alloc")
+            .and_then(|a| a.as_object())
+            .ok_or_else(|| anyhow::anyhow!("Missing 'alloc' in genesis"))?;
+
+        let mut balances = Vec::new();
+        for (eth_addr, obj) in alloc.iter() {
+            let balance_str = obj.get("balance").and_then(|b| b.as_str()).unwrap_or("0");
+            let balance = if balance_str.starts_with("0x") {
+                u128::from_str_radix(balance_str.trim_start_matches("0x"), 16)?
+            } else {
+                balance_str.parse::<u128>()?
+            };
+            let substrate_addr = &self.eth_to_substrate_address(eth_addr)?;
+            balances.push((substrate_addr.clone(), balance));
+        }
+        Ok(balances)
+    }
+
+    fn eth_to_substrate_address(&self, eth_addr: &str) -> anyhow::Result<String> {
+        let eth_bytes = hex::decode(eth_addr.trim_start_matches("0x"))?;
+        let mut padded = [0u8; 32];
+        padded[12..].copy_from_slice(&eth_bytes);
+        let account_id = AccountId32::from(padded);
```

Comment:
Are you sure this works? I assume they should be mapped addresses, meaning the filler bytes should be `EE` and not `00`? If they aren't mapped we have no way to find the corresponding private keys for signing the transactions.

Reply:
Yes sorry I should have pointed you to that code directly. But you got it correct. This is how it works.